Reclaim disk space story. (#1452)
pulp/pulpcore: docs/workflows/reclaim-disk-space.rst

.. _reclaim-disk-space:

Reclaim disk space for a list of repositories
---------------------------------------------

.. note::

>     Pulp reclaim disk space feature are in tech-preview and may change in backwards incompatible

Reviewer: ```suggestion
    Pulp's reclaim disk space feature is in tech-preview and may change in backwards incompatible
```

    ways in future releases.

> Pulp provides some feature for reclaiming disk space for:

Reviewer 2: "some" feels like a typo, maybe rwrite as "Pulp provides the ability to reclaim disk space for"?

>   * content that one no longer needs to serve but wants to keep it in the repo for historical

Reviewer 2: Same with "wants"

>   * content that one no longer needs to serve but would like to keep it in the repo for historical
>     purpose.

Reviewer: ```suggestion
  * content that one no longer needs to serve but would like to keep in the repo for historical reasons
```

  * repos that have on_demand download policy and one would want to clear out downloaded files for
    those repos. Usually, a repository with on_demand download policy will store artifacts locally
    after they have been requested, but there really isn't a way to have pulp delete the locally

>     stored files and free disk space if these packages are unlikely to be used again.

Reviewer 2: It would be a neat idea (for the future) to let a user configure this to be done automatically for old repository versions after a new one is created.

    $ http --body :24817$RECLAIM_TASK

    {
        "child_tasks": [],
        "created_resources": [],
        "error": null,
        "finished_at": "2021-07-16T15:36:20.650573Z",
        "logging_cid": "50d1721d205c40f69defb773e32a98ff",
        "name": "pulpcore.app.tasks.reclaim_space.reclaim_space",
        "parent_task": null,
        "progress_reports": [
           {
                "code": "reclaim-space.artifact",
                "done": 35,
                "message": "Reclaim disk space",
                "state": "completed",
                "suffix": null,
                "total": 35
            }
        ],
        "pulp_created": "2021-07-16T15:36:20.306845Z",
        "pulp_href": "/pulp/api/v3/tasks/20ee50bd-9392-4ebf-8f1d-d2f15474ebd6/",
        "reserved_resources_record": [
            "/pulp/api/v3/repositories/rpm/rpm/b3a6674d-181c-4e72-9412-7cbc747480ad/"
        ],
        "started_at": "2021-07-16T15:36:20.370245Z",
        "state": "completed",
        "task_group": null,
        "worker": "/pulp/api/v3/workers/ccc132c4-0445-4f55-b370-32c3662dce3c/"
     }

> As a result of this request disk space will be freed-up for:

Reviewer: ```suggestion
As a result of this request, disk space will be freed-up for:
```

  * artifacts only exclusive to that list of provided repos. No new versions will be created or

>     removed because content set will not change.

Reviewer 2: Rewriting this slightly:

> Artifacts that are exclusive to the list of provided repos. The content of the repository versions will not change and no repository versions will be created or removed.

>   * willingly, a ``epo_versions_keeplist`` option can be specified that will contain list of

Reviewer 2: Typo, "epo_versions_keeplist"

>   * willingly, a ``repo_versions_keeplist`` option can be specified that will contain list of repo

Reviewer: ```suggestion
  * an optional ``repo_versions_keeplist`` option can be specified that will contain list of repo
```

>     ``repo_version_hrefs`` which will be excluded from Artifact removal.

Reviewer 2: "repo_version_hrefs" doesn't look like it's a parameter and it's not a resource type, maybe it should be "repo version ``href``s?

  * by default, the task will remove only those Artifacts that have a corresponding RemoteArtifact.
  * for uploaded content that does not have RemoteArtifacts, a force flag should be specified.

>     This, however, will lead to a non-cosumable content afterwards.

Reviewer 2: Typo "cosumable"

I'm conflicted about this because `RemoteArtifact` is a very low-level detail and not something that we should expect users to know or care about.  But the risk of corruption makes it important, which feels like a major red flag.

Is a `force` flag a hard requirement?  Do we expect users to have a critical need to be able to delete content for which there is no longer any remote?

If it's not, I would be more comfortable leaving it out.  Or considering implementing something like [method 1d](https://hackmd.io/_3gsUVdyQwy50Nc5pMXN-g) to prevent the entire issue.

Author: automation hub needs a way to reclaim disk space for content that has been uploaded https://hackmd.io/e78_-T2rQRWvXaGjHT6Kyw#First-workflow-container-where-versions-are-kept-for-history

Reviewer 2: Then we really need to make sure that it fails as gracefully as possible if purged content is attempted to be downloaded.

Author: that's a good point, i will need to look into this and confirm

Reviewer 2: https://pulp.plan.io/issues/8305 is probably a good place to write the results

Author: I have not manually tested rpm plugin yet,  but skimmed through pulpcore code - plugins that use directly pulp's content app handler, should be able to gracefully handle this. Uploaded content will have artifact set to none as well it won't have any remoteartifacts so a 404 will just be raised https://github.com/pulp/pulpcore/blob/master/pulpcore/content/handler.py#L681.
Since pulp-container has a subclassed version of handler, i needed to modify couple of lines which led to this output:
```
$ podman pull localhost:24817/test/repo --tls-verify=false
Trying to pull localhost:24817/test/repo:latest...
Error: Error initializing source docker://localhost:24817/test/repo:latest: Error reading manifest latest in localhost:24817/test/repo: StatusCode: 404, 404: Not Found
(pulp) [vagrant@pulp3-source-fedora34 ~]$ 
```

tldr, i think we're fine just need to audit plugins that subclass the Handler.

Reviewer 2: Would you mind filing an issue to do that?

>   The task will trim Artifacts regardless of the download policy. Content app would be able to stream artifact if it is locally available, otherwise with the help of RemoteArtifact (if available) it would download artifact from the remote source.

Reviewer: Does trim here mean delete?

```suggestion
The task will clean up Artifacts regardless of the download policy. The content app will be able to
stream artifact if it is locally available, otherwise using the RemoteArtifact (if available), it
will download artifact from the remote source.
```

Author: yes

        "parent_task": null,
        "progress_reports": [
           {
                "code": "reclaim-space.artifact",
                "done": 35,
                "message": "Reclaim disk space",
                "state": "completed",
                "suffix": null,
                "total": 35
            }
        ],
        "pulp_created": "2021-07-16T15:36:20.306845Z",
        "pulp_href": "/pulp/api/v3/tasks/20ee50bd-9392-4ebf-8f1d-d2f15474ebd6/",
        "reserved_resources_record": [
            "/pulp/api/v3/repositories/rpm/rpm/b3a6674d-181c-4e72-9412-7cbc747480ad/"
        ],
        "started_at": "2021-07-16T15:36:20.370245Z",
        "state": "completed",
        "task_group": null,
        "worker": "/pulp/api/v3/workers/ccc132c4-0445-4f55-b370-32c3662dce3c/"
     }

As a result of this request, disk space will be freed-up for:

  * artifacts that are exclusive to the list of provided repos. The content of the repository
    versions will not change and no repository versions will be created or removed.
  * an optional ``repo_versions_keeplist`` parameter can be specified that will contain list of
    repo version hrefs which will be excluded from Artifact removal.

>   * by default, the task will remove only those Artifacts that have a corresponding RemoteArtifact.

Reviewer 2: I think the two bullet points about the `--force` option should be in a separate paragraph from the rest (or maybe in a "note" box) for visibility.  Alternatively we should use a big "warning" box such as the one on this page: https://docs.pulpproject.org/pulpcore/workflows/repairing-pulp.html

We can make it available but it should be clear that the user should never use this option without understanding the full implications.  I still think that needing to explain low-level details such as `RemoteArtifact` is a red flag and a usability problem.

We should also call out that if the content cannot be guaranteed to be / stay available at the remote source, either due to changes at the remote source or changes to the remote configuration.

Author: @dralley i tried to re-word couple of things is this is more user friendly and we don't talk about low-level stuff such ass 'remote artifact'. Let me know if that looks better to you and if i need to make any further adjustments.

  * for uploaded content that does not have RemoteArtifacts, a ``force`` flag should be specified.

>     This, however, will lead to a non-consumable content afterwards.

Reviewer 2: I still feel like we shouldn't need to explain `RemoteArtifacts` to the user -- and probably the `--force` option should have some kind of high-visibility note or warning box to call out the potential for things to go wrong.  We can make it available but the user should understand that it's something they really really shouldn't do unless they know what they are doing.

        ],
        "started_at": "2021-07-16T15:36:20.370245Z",
        "state": "completed",
        "task_group": null,
        "worker": "/pulp/api/v3/workers/ccc132c4-0445-4f55-b370-32c3662dce3c/"
     }

As a result of this request, disk space will be freed-up for:

  * artifacts that are exclusive to the list of provided repos. The content of the repository
    versions will not change and no repository versions will be created or removed.
  * an optional ``repo_versions_keeplist`` parameter can be specified that will contain list of
    repo version hrefs which will be excluded from the artifact removal.
  * by default, the task will remove only those artifacts from the content that were synchronized
    into Pulp.

.. warning::

    For uploaded content that, as a rule, does not have a remote source url, a ``force`` flag must
    be specified. This, however, will lead to a non-consumable content afterwards.
    
The task will clean up artifacts regardless of the download policy. The content app will be able
to stream artifact if it is locally available, otherwise it will attempt to redownload it from the

> known upstream urls. In case upstream stopped serving the corresponding file , Pulp won't be able to

Reviewer 2: Extra space before the comma

.. _reclaim-disk-space:

Reclaim disk space for a list of repositories
---------------------------------------------

.. note::

    Pulp's reclaim disk space feature is in tech-preview and may change in backwards incompatible
    ways in future releases.

Pulp provides the ability to reclaim disk space for:

  * content that one no longer needs to serve but would like to keep in the repo for historical
    reasons.

>   * repos that have on_demand download policy and one would want to clear out downloaded files for

Reviewer 2: s/have/were synced with

    $ http --body :24817$RECLAIM_TASK

    {
        "child_tasks": [],
        "created_resources": [],
        "error": null,
        "finished_at": "2021-07-16T15:36:20.650573Z",
        "logging_cid": "50d1721d205c40f69defb773e32a98ff",
        "name": "pulpcore.app.tasks.reclaim_space.reclaim_space",
        "parent_task": null,
        "progress_reports": [
           {
                "code": "reclaim-space.artifact",
                "done": 35,
                "message": "Reclaim disk space",
                "state": "completed",
                "suffix": null,
                "total": 35
            }
        ],
        "pulp_created": "2021-07-16T15:36:20.306845Z",
        "pulp_href": "/pulp/api/v3/tasks/20ee50bd-9392-4ebf-8f1d-d2f15474ebd6/",
        "reserved_resources_record": [
            "/pulp/api/v3/repositories/rpm/rpm/b3a6674d-181c-4e72-9412-7cbc747480ad/"
        ],
        "started_at": "2021-07-16T15:36:20.370245Z",
        "state": "completed",
        "task_group": null,
        "worker": "/pulp/api/v3/workers/ccc132c4-0445-4f55-b370-32c3662dce3c/"
     }

As a result of this request, disk space will be freed-up for:

> 

Reviewer 2: Maybe we should get rid of the list and switch it to paragraph form?  Points two and three don't match with the caption above.

  * artifacts that are exclusive to the list of provided repos. The content of the repository
    versions will not change and no repository versions will be created or removed.
  * an optional ``repo_versions_keeplist`` parameter can be specified that will contain list of
    repo version hrefs which will be excluded from the artifact removal.
  * by default, the task will remove only those artifacts from the content that were synchronized

>     into Pulp.

Reviewer 2: The meaning of this isn't totally clear, do you mean content that is already part of a repository version?

Author: i mean content that was synchronized in pulp. there are just 2 ways how to get content into pulp - you whether upload it or synchronize it.

Reviewer 2: I think it would make more sense to reverse it, i.e. "by default, the task will *not* remove artifacts from content that were uploaded directly into Pulp"

Author: sure, if this is more readable i'll update.

.. note::

   The task will clean up artifacts regardless of the download policy. The content app will be able
   to stream artifact if it is locally available, otherwise it will attempt to redownload it from

>    the known upstream urls. In case upstream stopped serving the corresponding file , Pulp won't be

Reviewer 2: Extra space before the comma

   able to download and serve it.

.. warning::

    For uploaded content that, as a rule, does not have a remote source url, a ``force`` flag must

>     be specified. This, however, will lead to a non-consumable content afterwards. 

Reviewer 2: @ipanova Do you think it would be reasonable to prevent content with no Artifact and no RemoteArtifact from being added to new repository versions via one of the validation hooks?

Author: don't we fall into the same category of the issue we have with the removal of remote, remoteartifacts are being removed as well? in case it was an on_demand content the ca will not have either artifact nor remoteartifact?

Author: what i am trying to say is whatether we'd decide to do with such content that does not have RA and artifact it should be done in a separate PR after figuring out what to do with https://hackmd.io/_3gsUVdyQwy50Nc5pMXN-g
but yeah, you've opened the pandora box here by asking this question :D

Reviewer 2: @ipanova I think it can be left for #8305, as long as we actually address #8305 soon :)

Author: #soon

Author: i have added a comment to that issue with your proposal of the validation hooks so it can be discussed whenever relevant.

Reviewer 2: Thanks!
